bears/jinja2/Jinja2Bear.py

```diff
@@ -0,0 +1,271 @@
+                    "`{0}` spaces on each side.".format(
+                        control_spacing),
+            file=filename,
+            line=line_number,
+            column=match_object.start(1) + 1,
+            end_line=line_number + 1,
+            end_column=match_object.end(1) + 1,
+            diffs=diff)
+
+    def run(self,
+            filename,
+            file,
+            variable_spacing: int=1,
+            control_spacing: int=1):
+        '''
+        Check Jinja2 templates for syntax, formatting and documentation issues.
+        The following aspects are being looked at:
+
+        * Variable spacing: Variable tags should be padded with one space on
+                            each side, like this: `{{ var_name }}`. This can
+                            be set to any number of spaces via the setting
+                            variable_spacing.
+                            Malformatted variable tags are detected and fixes
+                            suggested.
+        * Control spacing:  Like variable spacing, but for control blocks,
+                            i.e. if and for constructs. Looks at both start
+                            and end block.
+        * Control labels:   It is good practise to label the end of an if or
+                            for construct with a comment equal to the content
+                            of the start, like so:
+
+                            ```
+                            {% for x in y %}
+                            do something
+                            {% endfor %}{# for x in y %}
+                            ```
+
+                            Mising or differing labels are detected and
+                            fixes suggested.
+                            Constructs with start and end on the same line are
+                            being ignored.
+        * unbalanced blocks:   Each opening tag for a for or if construct must
+                               be closed by a corresponding end tag. An
+                               unbalanced number of opening and closing tags is
+                               invalid syntax and will be reported with MAJOR
+                               severity by the bear.
+
+
+        :param variable_spacing: The number of spaces a variable block
+                                 should be spaced with. Default is 1.
+        :param control_spacing:  The number of spaces a control block
+                                 should be spaced with. Default is 1.
+        '''
+        variable_regex = re.compile(
+            r"(?P<open>{{)(?P<content>.*?)(?P<close>}})")
+        control_start_regex = re.compile(
+            r"(?P<open>{%)(?P<content>\s*(for|if).*?)(?P<close>%})")
+        control_end_regex = re.compile(
+            r"(?P<open>{%)(?P<content>\s*end(for|if)\s*?)"
+            r"(?P<close>%})(?P<label>{#.*?#})?")
+
+        # Whenever a control construct starts a tuple of the
```

Comment:
Minor modifications for easier reading:

Whenever a control construct starts**,** a tuple of the expected end label and line number are put on this stack. Whenever an end tag is encounter**ed** the **last item added** is popped.